### ISSUE TITLE
Problem with texture alignment

### ISSUE THREAD

**Issue by fefogarcia** [YadiraF/DECA#29]
Hi folks, thank you for this mind-blowing model!!

I've been trying to use the reconstruction demo and while the geometry reconstruction works like magic, the texture is always misaligned, especially around the eyes. At first I thought it was an issue with the custom image I was using but then when I attempted one of the sample images provided I had the same result.

I then thought it might have been related to the low-poly, pre-displacement OBJ I was using, but then after fumbling around a bit with the normal map with no luck, I opened the high-detail OBJ to check the vertex colors and it turns out they were also misaligned in the same way

Have you guys seen this problem before? Maybe there's something in the MAT file I should be applying? Thank you in advance for any light you can shine on the issue!

![image](https://user-images.githubusercontent.com/9119493/113654415-223e2880-966e-11eb-8ff4-72b04e445431.png)

**Comment by vladyslavmos** [comment]
Same issue. Maybe someone know how to fix it?

**Comment by Dian-Yi** [comment]
Same issue. Maybe someone know how to fix it?

**Comment by mosvlad** [comment]
As I understand it, the problem is as follows.

For face-aligment this project use [FAN](https://github.com/1adrianb/face-alignment). And it's project is not so good for detecting narrowed eyes, for example :

![image](https://user-images.githubusercontent.com/31764930/115547471-f7c7ae80-a2ae-11eb-8b09-107ac4fe39a4.png)

![image](https://user-images.githubusercontent.com/31764930/115548038-ab30a300-a2af-11eb-845b-92ea788b3625.png)


Module for detecting facial landmarks [here](https://github.com/YadiraF/DECA/blob/97f253233c0f5c05b0e247b95e444008c7db8039/decalib/datasets/detectors.py)

Maybe someone know more robust project for detecting facial landmarks? Please share it, and we can implement it for this project.

**Comment by fefogarcia** [comment]
Hi @mosvlad, that is great insight! Now that you said that I took a closer look at the README example that uses this same photo from mine and the eyes are indeed misaligned – this whole time I had assumed they were correctly placed.

![Example from README](https://github.com/YadiraF/DECA/raw/master/Doc/images/soubhik.gif)

Re: facial alignment, I have used [MTCNN](https://github.com/ipazc/mtcnn) with good results when I've run Microsoft's [Deep 3D Facial Reconstruction](https://github.com/microsoft/Deep3DFaceReconstruction) model. Maybe it could become an option here to run MTCNN as an intermediate step to generate the landmarks before running DECA?

**Comment by mosvlad** [comment]
I think i can implement model from [Deep3DFaceReconstruction](https://github.com/microsoft/Deep3DFaceReconstruction) for detecting facial landmarks. Code look's simple for use.

[code for facial landmark detection from microsoft/Deep3DFaceReconstruction](https://github.com/microsoft/Deep3DFaceReconstruction/blob/5b131a3e67597da67409486e20db50007f48427d/preprocess_img.py#L139)

#Use model

	#detect 68 face landmarks for aligned images
	def get_68landmark(img,detector,sess):

		input_img = detector.get_tensor_by_name('input_imgs:0')
		lm = detector.get_tensor_by_name('landmark:0')

		landmark = sess.run(lm,feed_dict={input_img:img})
		landmark = np.reshape(landmark,[68,2])
		landmark = np.stack([landmark[:,1],223-landmark[:,0]],axis=1)

		return landmark

#Load model

		with tf.Graph().as_default() as graph, tf.device('/gpu:0'):
		lm_detector = load_graph(os.path.join('network','landmark68_detector.pb'))
		tf.import_graph_def(lm_detector,name='')
		sess = tf.InteractiveSession()

		for file in img_list:

			print(file)
			name = file.split('/')[-1].replace('.png','').replace('.jpg','')
			img,lm5p = load_img(file,file.replace('png','txt').replace('jpg','txt'))
			img_align,_,_ = align_img(img,lm5p,lm3D)  # [1,224,224,3] BGR image

			lm68p = get_68landmark(img_align,graph,sess)
			lm68p = lm68p.astype(np.float64)
			skin_mask = get_skinmask(img_align)

I will announce the results shortly. Thanks for the info on this project.

**Comment by Zju-George** [comment]
@mosvlad Is the MS repo for 68 landmarks detection better? It would be wonderful if you share some results of MS results.

**Comment by mosvlad** [comment]
@Zju-George
I think MS project may show a more accurate result for the narrowed eyes. In the near future I will try to share the results of comparison between [MS ](https://github.com/microsoft/Deep3DFaceReconstruction) and [FAN](https://github.com/1adrianb/face-alignment)

**Comment by mosvlad** [comment]
@Zju-George @fefogarcia Some results.

![image](https://user-images.githubusercontent.com/31764930/116140553-27781b80-a6e0-11eb-883e-e559766df022.png)
68 facial landmark model from MS repo - is [2D-and-3D-face-alignment](https://github.com/1adrianb/2D-and-3D-face-alignment) This repo is lua impementation of [FAN](https://github.com/1adrianb/face-alignment) :)

But I'm still trying to find robust and accuracy model for 68 facial landmark. Next step, i'll try to comparse FAN and [dlib](https://github.com/davisking/dlib)

It looks pretty accuracy

![image](https://user-images.githubusercontent.com/31764930/116141176-0237dd00-a6e1-11eb-8847-5a594340c936.png)

**Comment by mosvlad** [comment]
DLIB facial landmark detector is much better than FAN for frontal images. Here sample of code for detect landmarks with DLIB. 
I'm use GoogleColab for tests. 

For use download [shape_predictor_68_face_landmarks.dat](https://github.com/italojs/facial-landmarks-recognition/blob/master/shape_predictor_68_face_landmarks.dat)

    from imutils import face_utils
    import numpy as np
    import argparse
    import imutils
    import dlib
    import cv2
    
    from google.colab.patches import cv2_imshow
    
    
    detector = dlib.get_frontal_face_detector()
    predictor = dlib.shape_predictor("shape_predictor_68_face_landmarks.dat")
    
    image = cv2.imread("test.jpg")
    image = imutils.resize(image, width=500)
    gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    
    rects = detector(gray, 1)
    
    for (i, rect) in enumerate(rects):
	    shape = predictor(gray, rect)
	    shape = face_utils.shape_to_np(shape)
    
	    (x, y, w, h) = face_utils.rect_to_bb(rect)
	    cv2.rectangle(image, (x, y), (x + w, y + h), (0, 255, 0), 2)
    
	    cv2.putText(image, "Face #{}".format(i + 1), (x - 10, y - 10),
		    cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 2)
    
	    for (x, y) in shape:
		    cv2.circle(image, (x, y), 1, (0, 0, 255), -1)
    
    cv2_imshow(image)
    
    cv2.waitKey(0)


![image](https://user-images.githubusercontent.com/31764930/116205357-5628dd00-a746-11eb-99c3-7bbdb0dc06eb.png)


Next step implement DLIB landmarks for this project :)

**Comment by mosvlad** [comment]
Some interest result of my research about align eyes texture.

![image](https://user-images.githubusercontent.com/31764930/116436696-51068380-a855-11eb-9996-3bb10ce56248.png)
![image](https://user-images.githubusercontent.com/31764930/116437122-c83c1780-a855-11eb-97f5-b58dede4aad4.png)
![image](https://user-images.githubusercontent.com/31764930/116437082-b9edfb80-a855-11eb-83bc-4e80e41a1b7c.png)

I think I am close to a solution. I will give more information about this work soon.

**Comment by timlod** [comment]
This project doesn't use landmark detection for anything other than getting a bounding box for the face - thus I think getting better landmark detection is not as important as you may think. You can investigate this in `datasets.py` - see how the box is used to compute a center and size in `bbox2point`, and then used to estimate and compute a transform on the image.

Internally, DECA computes landmarks itself, which is what is shown in the output. For training, the paper states that they used FAN to create ground-truth landmarks for all training data, so bad results in the trained model probably arise from bad landmarks used as ground truth in the training set.

**Comment by mosvlad** [comment]
@timlod 
Yes, everything is exactly as you wrote it. At first I thought that the face points are used directly to generate the model, but after more detailed analysis I realized that the points are not used directly. But 2D points are used when projecting the texture onto the finished model, at this stage you can use the facial point detector to correct the input image and the eyes were projected to the right place.

Here is reult for sample image:

![image](https://user-images.githubusercontent.com/31764930/116883316-28540480-ac2e-11eb-8641-224d2d6dd141.png)
![image](https://user-images.githubusercontent.com/31764930/116883372-3570f380-ac2e-11eb-88dd-2ab034fe1ab9.png)

**Comment by mjoach** [comment]
@mosvlad, just run into the same issue and confirmed your findings, thanks for sharing your experiences. But how did you correct UV texture with externally provided landmarks above? Based on my understanding of the source, DECA itself doesn't use landmarks when unwrapping input to the UV texture so there is no single point to feed it proper landmarks. Am I wrong on this one? Or are you applying some corrective stretching to the input image/unwrapped texture?

**Comment by timlod** [comment]
@mosvlad I'd also be interested in how you used custom landmarks for the texture projection!

**Comment by birdflies** [comment]
> @mosvlad Is there a chance that you could share the code/modifications you used to get the final result? I came across the same issue of misaligned eyes and your solution could really help me with a research project I’m working on.

Hi, Do u solve this problem~

**Comment by MokzZ** [comment]
@birdflies I think I figured out how @mosvlad solved it. Or at least I found a similar solution. In `deca.py`, in the `def decode()` section, `landmarks2d` (line 174) is the Torch tensor containing the landmarks used to show the visualization landmarks in the generated `..._vis.jpg` files. These landmarks can be used to create bounding boxes around the eyes. Then implementing the DLIB landmark detector and using the landmarks to create different bounding boxes around the eyes, you can copy the texture of the eyes found by DLIB to the location where DECA thinks the eyes are. I use opencv to do the copying. You can then overwrite the image loaded into the `visdict` (line 245), which is the image used to create the model texture. It is not the cleanest solution, but it does the job for me at least.

**Comment by birdflies** [comment]
> @birdflies I think I figured out how @mosvlad solved it. Or at least I found a similar solution. In `deca.py`, in the `def decode()` section, `landmarks2d` (line 174) is the Torch tensor containing the landmarks used to show the visualization landmarks in the generated `..._vis.jpg` files. These landmarks can be used to create bounding boxes around the eyes. Then implementing the DLIB landmark detector and using the landmarks to create different bounding boxes around the eyes, you can copy the texture of the eyes found by DLIB to the location where DECA thinks the eyes are. I use opencv to do the copying. You can then overwrite the image loaded into the `visdict` (line 245), which is the image used to create the model texture. It is not the cleanest solution, but it does the job for me at least.

Thanks, do you solve problems : "the results are all black except for the face?"?

**Comment by MokzZ** [comment]
@birdflies I haven't come across that problem myself. I use the `--useTex` argument while running and it succesfully generates a full uv texture map.

**Comment by vedantdere** [comment]
> @birdflies I haven't come across that problem myself. I use the `--useTex` argument while running and it succesfully generates a full uv texture map.

Please can you share the model which you have used to generate full texture map

**Comment by VictoryLoveJessica** [comment]
I think it can be solved by new face detection
![DECA_new](https://user-images.githubusercontent.com/73265430/182338224-4cdcb871-9cfe-4b27-8dbe-b7e1573d37ee.png)

**Comment by emlcpfx** [comment]
> I think it can be solved by new face detection ![DECA_new](https://user-images.githubusercontent.com/73265430/182338224-4cdcb871-9cfe-4b27-8dbe-b7e1573d37ee.png)

Can you please elaborate? I'd love to implement this with EMOCA.

**Comment by emlcpfx** [comment]
> @timlod Yes, everything is exactly as you wrote it. At first I thought that the face points are used directly to generate the model, but after more detailed analysis I realized that the points are not used directly. But 2D points are used when projecting the texture onto the finished model, at this stage you can use the facial point detector to correct the input image and the eyes were projected to the right place.
> 
> Here is reult for sample image:
> 
> ![image](https://user-images.githubusercontent.com/31764930/116883316-28540480-ac2e-11eb-8641-224d2d6dd141.png) ![image](https://user-images.githubusercontent.com/31764930/116883372-3570f380-ac2e-11eb-88dd-2ab034fe1ab9.png)

Hi Vlad, Can you please elaborate on your findings? How did you get this to work?

**Comment by emlcpfx** [comment]
@yfeng95 Can you please weigh in on the technique that @mosvlad is proposing?

**Comment by emlcpfx** [comment]
I found a hacky fix for how this eye alignment was affecting our work. Check it out here:
https://github.com/radekd91/emoca/issues/61#issuecomment-1683007725

There are a lot of very smart people on this board, maybe you can weigh in on how I could do this better.
